core/trino-main/src/main/java/io/trino/sql/planner/rowpattern/IrAnchorPattern.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.rowpattern;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.trino.sql.tree.AnchorPattern;
+
+import java.util.Objects;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class IrAnchorPattern
+        extends IrRowPattern
+{
+    public enum Type
+    {
+        PARTITION_START,
+        PARTITION_END;
+
+        public static Type from(AnchorPattern.Type type)
+        {
+            switch (type) {
+                case PARTITION_START:
+                    return PARTITION_START;
+                case PARTITION_END:
+                    return PARTITION_END;
+            }
+            throw new IllegalArgumentException("unsupported anchor type: " + type);
+        }
+    }
+
+    private final Type type;
+
+    @JsonCreator
+    public IrAnchorPattern(Type type)
+    {
+        this.type = requireNonNull(type, "type is null");
+    }
+
+    @JsonProperty
+    public Type getType()
+    {
+        return type;
+    }
+
+    @Override
+    public <R, C> R accept(IrRowPatternVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitIrAnchorPattern(this, context);
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if ((obj == null) || (getClass() != obj.getClass())) {
+            return false;
+        }
+        IrAnchorPattern o = (IrAnchorPattern) obj;
+        return Objects.equals(type, o.type);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return 0;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
```

Comment:
The toString() in all these classes could be improved so that the output is closer to what a pattern would look like. That would make them much easier to read and interpret when debugging.